### ISSUE TITLE
[New Feature] (?) Global `after` configuration?

### ISSUE THREAD

**Issue by vi013t** [zaldih/themery.nvim#23]
I have some bits of colorscheme/highlight changes:

```lua
vim.api.nvim_set_hl(0, "NeoTreeGitUntracked", { fg = "#88FF88" })
vim.cmd("hi CursorLineNr gui=bold")
```
...that I would like to apply no matter what my theme is. The only way I see to do this so far is to copy and paste it into the `after` key for every individual theme. I was wondering if there's a more convenient way to run this after any theme is applied? Like a global `after` option or even a `ThemeChanged` event of sorts that could be listened for? Normally I had these lines just in my `config` for some plugins, but changing the theme with Themery seems to (logically) override these highlight changes.

Thank you!

**Comment by zaldih** [comment]
Thank you for opening the issue!

You are right! It didn't occur to me before and it makes sense that there are hooks to capture events like changing themes (also, to allow other plugins to make use of it).

As soon as I have some time I'll get on this. It's a simple thing to implement.

**Comment by vi013t** [comment]
Thank you!

On the topic of `before` and `after`, is there any reason why these are strings of Lua code instead of lua functions? I see you're using `load` to call them ([here](https://github.com/zaldih/themery.nvim/blob/f745a49d9c103babde35e1111b91faa6bc591d5c/lua/themery/controller.lua#L31)); Is there any reason to prefer this over something like:

```lua
if theme.before then
    local success, err = pcall(theme.before)
    if not success then
        print("Themery error: " .. err)
        return false
    end
end
```

This is, in my experience, the more traditional way of doing this, and it has the advantage that the user can write a Lua function with proper intellisense/linting/autocomplete, instead of a string of code. Is there any reason why `before` and `after` are currently implemented as strings?

Sorry this is a bit off-topic, maybe it should be its own issue; I just thought it was worth bringing up on the topic.

**Comment by zaldih** [comment]
Don't worry, these questions are great because I am far from being an expert in lua and I have a lot to learn! Besides, it comes to the point.

The main and only reason for defining it as a string is that it was the only way I managed to make it work to serialize the functions and be able to use them later when loading each colorscheme.

Until recently, the after and before code had to be saved flat in a file for neovim to load it at startup, so I needed to be able to get the function text. By defining them as a function, lua generated its intermediate code which cannot be serialised.

In short, they are defined as text so that they can be stored in plain text.

Of course there may be better ideas. Now, for example, functions are no longer stored in another file, since it is the plugin that takes care of them, and what you propose makes a lot of sense. I will think about it!

---
## References

- https://github.com/zaldih/themery.nvim/tree/v1.0.0?tab=readme-ov-file#persistence
- #11

**Comment by LmanTW** [comment]
> Don't worry, these questions are great because I am far from being an expert in lua and I have a lot to learn! Besides, it comes to the point.
> 
> The main and only reason for defining it as a string is that it was the only way I managed to make it work to serialize the functions and be able to use them later when loading each colorscheme.
> 
> Until recently, the after and before code had to be saved flat in a file for neovim to load it at startup, so I needed to be able to get the function text. By defining them as a function, lua generated its intermediate code which cannot be serialised.
> 
> In short, they are defined as text so that they can be stored in plain text.
> 
> Of course there may be better ideas. Now, for example, functions are no longer stored in another file, since it is the plugin that takes care of them, and what you propose makes a lot of sense. I will think about it!
> ## References
> 
>     * https://github.com/zaldih/themery.nvim/tree/v1.0.0?tab=readme-ov-file#persistence
> 
>     * [Change in Approach: Simplifying Persistence for Easier Adoption - Share Your Thoughts! #11](https://github.com/zaldih/themery.nvim/issues/11)

Also the reason why it need to save the function to plain text is because it loads the colorscheme before the setup function is called. So it actually have no access to the config when loading the colorscheme, there for the functions need to be store in the state file.

**Comment by zaldih** [comment]
Issue resolved thank to @LmanTW 🎉 .

@vi013t I'm going to give a second thought to what you said about defining the code as a function instead of text. If you want, you can open an issue in case someone wants to tackle it.

**Comment by LmanTW** [comment]
I just did some researching, I think the simplest solution is to use `string.dump()` to save the functions to a binary string, and use `load()` to load the function back. I don't know if we can just save it into the JSON file though, it may break the encoding.
